How to use GDB to debug MPI program (#1)
ImperialCollegeLondon/icl-rcs-user-guide: docs/hpc/advanced/debugging_mpi_programs/gdb_mpi_debug.md

6. **Conditional Breakpoints**: Developers can set breakpoints that trigger only when certain conditions are met, allowing for more precise debugging in complex scenarios.

7. **Core File Analysis**: GDB can analyze core dump files generated when a program crashes, providing insight into the state of the program at the time of the crash.

Again, we do not aim to teach the basics of GDB in this page and we assume that the user is familiar with the basic usage. You can keep the [GDB cheat sheet](https://darkdust.net/files/GDB%20Cheat%20Sheet.pdf) open which will help you to learn its usage faster and it will also be useful in this tutorial.

## Debugging MPI Programs with GDB

The process of debugging MPI programs with GDB is slightly more complex than debugging serial programs due to the parallel nature of MPI applications. When debugging an MPI program, you are dealing with multiple processes running concurrently, each with its own memory space and execution context. GDB provides features to debug MPI programs, but it requires additional steps and considerations to effectively debug parallel applications.

> Consider that you have an MPI program as shown below. The sample program given below is simple MPI program that performs addition and multiplication on each rank and prints the result. The first argument to the multiplication functions is the rank of the process and the second argument is the size of the communicator (this is actually the bug  which is there in the code below and our task is to use gdb on such programs to find the issue). We have create this  simple  program using functions so that you can step inside while debugging. Let us assume that the file is named `mpi_debug.cpp`.

Reviewer: ```suggestion
Consider that you have an MPI program as shown below. The sample program given below is simple MPI program that performs addition and multiplication on each rank and prints the result. The first argument to the multiplication functions is the rank of the process and the second argument is the size of the communicator (this is actually the bug  which is there in the code below and our task is to use gdb on such programs to find the issue). We will create a simple  program using functions so that you can step inside while debugging. Let us assume that the file is named `mpi_debug.cpp`.
```

 
```c++
#include <iostream>
#include <mpi.h>

// Function for multiplication
int multiplication(int a, int b, int rank, const char* hostname) {
    std::cout << "Rank " << rank << " on host " << hostname << " performing multiplication with arguments: " << a << " and " << b << std::endl;
    int result = a * b;
    std::cout << "Result of multiplication: " << result << std::endl;
    return result;
}

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);
    

>     int rank, size;

Reviewer: ```suggestion
    int rank, size, hostnamelen;
```

Author: I will add the code as suggested.

    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

> 

Reviewer: ```suggestion
    MPI_Get_processor_name (hostname, &hostnamelen);
```

Author: I will use the correct code and push it again so that we can merge it.

    return getpid();
}

// Insert this code right after the  MPI initialization routines (though not a mandatory requirement 
// to add there only). Please make a judgement based on your code.
// Retrieve process ID and hostname
pid_t pid = get_process_id();

volatile int i = 0;
while (0 == i)
{
	std::cout << "My rank = " << rank << " PID = " << pid << " running on Host = " << hostname << " in sleep " << std::endl;
	sleep(5);
}
```

We used the header file `unistd.h` which defines the function `getpid()`.

Based on the above output, we can easily figure out the process id of each rank and its host. Depending on where the process is running, you may have to ssh to that host and use GDB there.

Please note that the above logic is forcing all ranks to sleep infinitely unless we attach GDB to it and ask it to run. You can also change the above logic to a specific rank only if you are sure that this is the rank causing issues.

7) After adding the above code, we compile again and run our code. We will redirect the standard output and input to a file called `output.log` and run the process in the background as shown below.

```bash
mpirun -np 2 ./mpi_debug > output.log 2>&1 &
```

> where `2>&1` is used to redirect the standard error to the standard output and `&` is used to run the process in the background.

Reviewer: ```suggestion
where `2>&1` is used to redirect the standard error to the standard output and `&` is used to run the process in the background. When this is run a PID number will be returned which will be needed to attach gdb.
```

#2  0x00000000004013d6 in main (argc=1, argv=0x7ffe1214e358) at 5b_problem_mpi.cpp:36
```

Thus we can see that we are two levels deep in the main function. We can move up two levels by using the up command twice or we can directly jump to the main function by using the following command.

```bash
(gdb) up 2

# This will give the output
#2  0x00000000004013d6 in main (argc=1, argv=0x7ffe1214e358) at 5b_problem_mpi.cpp:36
 sleep(5);
```

Thus we are now in main function and we can change the value of i here.

10) We change the value of `i` to `1` by using the following command.

```bash
(gdb) set var i = 1

# Now we print i to see its value.
```bash
(gdb) print i
$2 = 1
```

As we can see that the value has been changed.

11) Next, we set a breakpoint at the multiplication function and run the program. We do this by using the following commands.

```bash
(gdb) break multiplication
(gdb) continue
```

12) The program at rank 1 now runs till the breakpoint which is the start of the function definition in our case. Inside this function, we can now print the values of both the arguments to see what was the issue.

```bash
(gdb) print a
$3 = 1
(gdb) print b
$4 = 1
```

> And, finally we can see that the issue is with the value of `b` which is `1` and not `6`. This is the bug in the code and we can fix it now.

Reviewer: ```suggestion
And, finally we can see that the issue is with the value of `b` which is `1` and not `2`. This is the bug in the code and we can fix it now by going back and setting num2 = size. Again this is a toy example but helps demonstrate the how to find a bug in a code base using a debugger.
```
